shared/src/TwinCollectionArray.cs

```diff
@@ -0,0 +1,77 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Azure.Devices.Shared
+{
+    /// <summary>
+    /// Represents a property array in <see cref="TwinCollection"/>
+    /// </summary>
+    public class TwinCollectionArray : JArray
+    {
+        private readonly JObject _metadata;
+
+        internal TwinCollectionArray(JArray jArray, JObject metadata)
+            : base(jArray)
+        {
+            _metadata = metadata;
+        }
+
+        /// <summary>
+        /// Gets the value for the given property name
+        /// </summary>
+        /// <param name="propertyName">Property Name to lookup</param>
+        /// <returns>Property value, if present</returns>
+        public dynamic this[string propertyName]
+        {
+            get
+            {
+                if (propertyName == TwinCollection.MetadataName)
+                {
+                    return GetMetadata();
+                }
+
+                if (propertyName == TwinCollection.LastUpdatedName)
+                {
+                    return GetLastUpdated();
+                }
+
+                if (propertyName == TwinCollection.LastUpdatedVersionName)
+                {
+                    return GetLastUpdatedVersion();
+                }
+
+                throw new ArgumentException($"{nameof(TwinCollectionArray)} does not contain a definition for '{propertyName}'.");
+            }
+        }
+
+        /// <summary>
+        /// Gets the Metadata for this property
+        /// </summary>
+        /// <returns>Metadata instance representing the metadata for this property</returns>
+        public Metadata GetMetadata()
```

Comment:
this, and below: they are common for both TwinCollectionArray and TwinCollectionValue; but with multiple inheritance not supported in c#, I've kept them repeated. I could define an interface with implementation, and use that, but it seemed too much of work for this.
What do you think?

Reply:
I don't feel too much code is duplicated. If you want to pursue a common base class, that'd be fine, but I don't expect it.

Reply:
I am going to leave it as-is. It isn't a whole lot of repeated code and I don't expect it to grow in the future either.